### ISSUE TITLE
Idea: User Plugin Submission Website

### ISSUE THREAD

**Issue by eMorph** [WayfireWM/wayfire#974]
In my opinion, one of the factors which led to the downfall of Compiz was that there was absolutely zero formal documentation for creating plugins and a complete lack of a central area for people to post their custom plugins to the public. Had these two things existed, Compiz 0.9 might have been the predominant window manager today.

Considering that Wayfire is a similar compositor to Compiz, following a similar model of externally-compiled plugins already, as well as that Wayfire plugins can be installed without needing to recompile the entire compositor, it seems to me to be a no-brainer that there should be a central hub on the Web for people to submit and release the plugins they've made. What I am considering is a website like Pling's theme repositories or extensions.gnome.org, which would include a search function, an easy way to upload plugins, and full API documentation. Perhaps even this could be an extension of Wayfire's current website, with new sub-pages being called Docs and Plugins.

This would make it easier for people to find new functionality for their instance of Wayfire, and would attract more code- or customization-minded people to the platform. With luck, the community this site would create would end up producing community plugins which provide libraries for other community plugins!

As for API documentation, I'd figured since there was already a tutorial for writing plugins, there might as well be simple documentation for each part of the Wayfire API, based on the tutorials already written. It worked with Awesome, XMonad, and Qtile, so it could work with a similarly extensible window manager.

**Comment by AdrianVovk** [comment]
Are you proposing people download pre-compiled binaries that other users uploaded to the site? That's really not a good idea for security, and it isn't really feasible for all the different distros (for example, where's the libc you're linking against? Where's libwayfire? What version of wf-config can you expect to be installed?)

If anything, maybe there could be some built in support in WCM or a command-line tool to download a plugin's source code, compile it, and enable it

**Comment by ammen99** [comment]
> Are you proposing people download pre-compiled binaries that other users uploaded to the site? That's really not a good idea for security, and it isn't really feasible for all the different distros (for example, where's the libc you're linking against? Where's libwayfire? What version of wf-config can you expect to be installed?)

That's a good point, pre-compiled binaries can never be an option. There are a ton of different distros, we can't support them all (with binaries).

> it seems to me to be a no-brainer that there should be a central hub on the Web for people to submit and release the plugins they've made

So far, there haven't been too many plugins, and I have started this wiki page as a way to discover them: https://github.com/WayfireWM/wayfire/wiki/Related-projects

That being said, it is far from perfect. We could maybe make a separate repository where we just use the wiki (or the repo itself) for showcasing plugins? I definitely do not have the skills and the motivation for a full-blown site like gnome's extensions (if someone else steps in, well, I don't see anything against that).

> As for API documentation, I'd figured since there was already a tutorial for writing plugins, there might as well be simple documentation for each part of the Wayfire API, based on the tutorials already written. It worked with Awesome, XMonad, and Qtile, so it could work with a similarly extensible window manager.

Wayfire's API is likely bigger than any of these window managers (maybe not in terms of individual functions and toggles, but in terms of the overall scope). As you have noticed, I have also started a series of smallish plugins with a more detailed explanation, in order to make it easier to get started. Also, I am trying to document every API function in the header source code.

So, it is not very clear to me, what exactly would you like to see in this documentation - perhaps elaborate a bit? :)

**Comment by VaZark** [comment]
Just jumping in to add that the Ulauncher launcher has a plugin installer tab in the app settings where the user has to enter the path to a plugin's repo so the application can download the source code and compile it. It also has a [website](https://ext.ulauncher.io/) that acts as a central hub where the approved/accepted plugins are listed along with it the repo path required by the plugin installer.

A similar web page that can serve as a central hub for third-party plugins while wcm or a separate minimalist app can handle plugin installation and removal.

**Comment by soreau** [comment]
Unfortunately, wayfire does not look for plugins in a user owned location like compiz does. Since the plugin .so and .xml paths are typically owned by root, wcm would have to install as root. If it had ability to uninstall as well, this would also require root and a way to track installed files. When plugins are being developed, many times the files will change name and locations, so there might be a chance to leave the installation prefix cluttered with files that may or may not have adverse effects down the road. If wayfire did look to load plugins from a user owned location, many of these issues could be solved.

The other problem I see is that a plugin may have some dependency and if it's not already installed, the best wcm could do is spit the failed output into a gtk window, so the user might have a clue why it failed to build. On second thought, maybe this isn't the biggest issue but perhaps this entire functionality should only be built when passing a build-time option.

**Comment by damianatorrpm** [comment]
IMHO the best option is that plugins are installed via packages from your distribution.

**Comment by travankor** [comment]
In case you haven't seen this repo yet: https://github.com/WayfireWM/wayfire-plugins-extra

Lots of extra plugins in here. I think random 3rd-party plugins will be a major annoyance, and right now Wayfire is still a (relatively) small project.

**Comment by soreau** [comment]
Yes, the truth is that even though writing wayfire plugins is made much easier because of the flexible API, great care must still be take to make sure the plugin doesn't cause visual bugs, leak memory or, of course, crash the compositor. However, I do think wayfire should look in a user owned location for plugin files first, before loading plugins from system installation location. For example, there could be ~/.wayfire/plugins/ for .so files and ~/.wayfire/metadata/ for .xml files. This would allow a user to write their own plugins and install the result to those locations as user. So if a user wants to make a change to a plugin that will not be accepted upstream, this will provide an easier way to do that without cluttering the install prefix.

**Comment by AdrianVovk** [comment]
Or, better yet, install it into where plugins exist in the prefix, but s@/usr@~/.local@

And, IIRC, you can put a direct path to a plugin .so in the config file and it'll load it

**Comment by eMorph** [comment]
In regards to the request for clarification as for documentation, I can clear up a few things. Due to Wayfire's plugins only being written in a combination of C++ and XML officially, it makes some sense to restrict the documentation to these two languages. In addition, it would also thus make sense to have the documentation be divided by language, clearing up certain things. The XML documentation would be relatively straightforward, simply detailing how plugin details should be written, and what every key does. However, the C++ documentation is slightly more complicated.

Because of the subject matter, that being detailing the specifics of a library written for C++ extensions, I propose that the documentation for this aspect would be organized in the same manner as the Apple Developer Documentation is for the Swift core libraries, and as Oracles documentation is for their various Java utilities. Essentially, this would mean that functions, types, and parameters are sorted as a tree, and the tree can be navigated via menus and links similar to a table of contents. This documentation does not need to be extensively detailed, as its only purpose would be to inform users and developers of the tools Wayfire uses.

A documentation page for a class in the Wayfire/wcm libraries would, given this format, include:

- Its name
- Its initial declaration (e.g. `public static class sampleClass_t`)
- A description of its role
- Names and short descriptions of its main methods and properties
- A list of every method and property it contains, with each entry having a link to its specific page

You don't need to describe in this documentation anything not provided specifically by Wayfire, including OpenGL, Cairo, libinput, etc., since that's the job of other people. Unless there's a Wayfire-specific wrapper or use, you can leave it out if Wayfire doesn't provide it.

**Comment by ammen99** [comment]
> Because of the subject matter, that being detailing the specifics of a library written for C++ extensions, I propose that the documentation for this aspect would be organized in the same manner as the Apple Developer Documentation is for the Swift core libraries, and as Oracles documentation is for their various Java utilities. Essentially, this would mean that functions, types, and parameters are sorted as a tree, and the tree can be navigated via menus and links similar to a table of contents. This documentation does not need to be extensively detailed, as its only purpose would be to inform users and developers of the tools Wayfire uses.

Sounds to me like you want the documentation generated by Doxygen :) For example, this is what Doxygen generates for the workspace-manager class: 

![slurped](https://user-images.githubusercontent.com/6287487/104708462-0afe3780-571e-11eb-9a8c-97a9680125e2.png)

If you generate these docs, you'll easily notice that many parts of the Wayfire API are underdocumented. This is definitely a direction Wayfire can(and should) improve in.

**Comment by eMorph** [comment]
> If you generate these docs, you'll easily notice that many parts of the Wayfire API are underdocumented. This is definitely a direction Wayfire can(and should) improve in.

Doxygen is a decent tool for a C/++ API such as this, but further description within the documentation itself, beyond simple descriptions already written, seems like a decent idea, seeing as the entire point of documentation would be to further explain the library beyond what is already written within the library itself. With a documentation system, explanation is not only more accessible to the average user (who could, with this knowledge of Wayfire's extensibility, become a part of Wayfire's customization community), but is also much easier to manage, edit, and extend. Without organized documentation, any explanation and instruction would have to be provided through comments, small amounts of information on wiki pages, and various tutorials, making it both much harder to find and search through and more difficult to maintain and update. Centralization allows developers to more easily document new objects, extend the current documentation, and update the explanation to reflect changes. In my eyes, it's a win-win for everyone involved.

**Comment by ammen99** [comment]
@eMorph Ah, I think I finally understand, you'd probably like docs like for example these here : https://www.oreilly.com/library/view/linux-device-drivers/0596005903/ch17.html

I'll be honest: such an effort will take a lot of time, and I'll probably not be writing them anytime soon. Of course, if anyone is up to the task, I can provide any info that this person can't figure out themselves ;)

**Comment by eMorph** [comment]
Sorta like that, but with a bit less elaboration. I think really just a description and basic usage is fine—nobody reads text walls.

**Comment by ammen99** [comment]
My brain dump from today, might serve as a good starting point if someone decides to write this stuff:

https://github.com/WayfireWM/wayfire/wiki/%5BWIP%5D-What-does-a-compositor-do%3F
https://github.com/WayfireWM/wayfire/wiki/%5BWIP%5D-Design-overview
